### PR TITLE
[NPM] [BUG] Supporting multiple values under label selector MatchExpr

### PR DESCRIPTION
**Pull request by vakalapa**
This PR is fix the bug of supporting multiple values in a LabelSelector match expressions. 
```yaml
 - key: pod
    operator: In
    values:
      - b
      - c
```
Today NPM considers the values of the above selector to be b AND c, but it should be b OR c

NPM now follows different design for podSelector and nameSpaceSelector. 

For PodSelector, NPM has ipsets with immutable members like ips or ip,proto:port. Ipset allows creating 2nd level ipset ype list:set to be created with pod label ipsets as members and NPM will  use this 2nd level ipset as the matchset criteria in rules.

For NameSpaceSelector, NPM uses ipsets with mutable members like other ipsets. Ipset does not allow adding list:set as members to list:set ipset, so 2nd level ipset cannot be used in case of nameSelectors. For this reason NPM will create combinations of nameSpaceSelectors and add individual rule for each combination.

### CODE REVIEW

File: npm/ipsm/ipsm.go [hunk 1]
```diff
@@ -70,6 +87,21 @@ func (ipsMgr *IpsetManager) Exists(listName string, setName string, kind string)
 	return true
 }
 
+// IpSetReferIncOrDec checks if an element exists in setMap/listMap and then increases or decreases tis refer count.
```
Comment (JungukCho) [inline npm/ipsm/ipsm.go:90]: tis refer count -> referCount in the Ipset

File: npm/ipsm/ipsm.go [hunk 2]
```diff
@@ -36,11 +44,20 @@ type Ipset struct {
 	referCount int
 }
 
+func (ipset *Ipset) IncReferCount() {
```
Comment (JungukCho) [inline npm/ipsm/ipsm.go:47]: I think it is better not to expose this function to other packages. IncReferCount() -> incReferCount

File: npm/ipsm/ipsm.go [hunk 3]
```diff
@@ -36,11 +44,20 @@ type Ipset struct {
...
+	ipset.referCount++
+}
+
+func (ipset *Ipset) DecReferCount() {
```
Comment (JungukCho) [inline npm/ipsm/ipsm.go:51]: I think it is better not to expose this function to other packages. DecReferCount() -> decReferCount

File: npm/ipsm/ipsm.go [hunk 4]
```diff
@@ -70,6 +87,21 @@ func (ipsMgr *IpsetManager) Exists(listName string, setName string, kind string)
 	return true
 }
 
+// IpSetReferIncOrDec checks if an element exists in setMap/listMap and then increases or decreases tis refer count.
+func (ipsMgr *IpsetManager) IpSetReferIncOrDec(ipsetName string, kind string, countOperation ReferCountOperation) {
```
Comment (JungukCho) [inline npm/ipsm/ipsm.go:91]: It is better not to expose this function to other packages. IpSetReferIncOrDec() -> ipSetReferIncOrDec.

May Improve readability by making separate functions for increasing and decreasing reference count instead of using flag if possible, but, do not have strong opinion.

File: npm/networkPolicyController.go [hunk 5]
```diff
@@ -351,10 +351,24 @@ func (c *networkPolicyController) syncAddAndUpdateNetPol(netPolObj *networkingv1
...
+	}
+	// Then NPM will add members to the above list, this is to avoid members being added
+	// to lists before they are created.
+	for listKey, listLabelsMembers := range lists {
+		for _, listMember := range listLabelsMembers {
+			if err = ipsMgr.AddToList(listKey, listMember); err != nil {
+				return fmt.Errorf("[syncAddAndUpdateNetPol] Error: Adding ipset member %s to ipset list %s with err: %v", listMember, listKey, err)
+			}
 		}
+		ipsMgr.IpSetReferIncOrDec(listKey, util.IpsetSetListFlag, ipsm.IncrementOp)
```
Comment (JungukCho) [inline npm/networkPolicyController.go:371]: better to include this call in AddToList(). So, we do not need to expose those functions to other packages and encapsulate information in ipsMgr.

File: npm/networkPolicyController.go [hunk 6]
```diff
@@ -351,10 +351,24 @@ func (c *networkPolicyController) syncAddAndUpdateNetPol(netPolObj *networkingv1
...
-		if err = ipsMgr.CreateList(list); err != nil {
-			return fmt.Errorf("[syncAddAndUpdateNetPol] Error: creating ipset list %s with err: %v", list, err)
+
+	// lists is a map with list name and members as value
+	// NPM will create the list first and increments the refer count
+	for listKey, _ := range lists {
+		if err = ipsMgr.CreateList(listKey); err != nil {
+			return fmt.Errorf("[syncAddAndUpdateNetPol] Error: creating ipset list %s with err: %v", listKey, err)
+		}
+		ipsMgr.IpSetReferIncOrDec(listKey, util.IpsetSetListFlag, ipsm.IncrementOp)
```
Comment (JungukCho) [inline npm/networkPolicyController.go:361]: better to include this call in CreateList(). So, we do not need to expose this function to other packages and encapsulate information in ipsMgr.

File: npm/parseSelector.go [hunk 7]
```diff
@@ -14,18 +14,18 @@ import (
 // An ReqHeap is a min-heap of labelSelectorRequirements.
 type ReqHeap []metav1.LabelSelectorRequirement
 
-func (h ReqHeap) Len() int {
```
Comment (JungukCho) [inline npm/parseSelector.go:17]: Where does NPM use heap?

File: npm/parseSelector.go [hunk 8]
```diff
@@ -121,73 +121,170 @@ func sortSelector(selector *metav1.LabelSelector) {
 	selector.MatchExpressions = sortedReqs
 }
 
+// getSetNameForMultiValueSelector takes in label with multiple values without operator
+// and returns a new 2nd level ipset name
+func getSetNameForMultiValueSelector(key string, vals []string) string {
+	rtStr := key
```
Comment (JungukCho) [inline npm/parseSelector.go:127]: What is "rtStr"?

File: npm/parseSelector.go [hunk 9]
```diff
@@ -121,73 +121,170 @@ func sortSelector(selector *metav1.LabelSelector) {
...
-func parseSelector(selector *metav1.LabelSelector) ([]string, []string, []string) {
+// flattenNameSpaceSelector will help flatten multiple NameSpace selector match Expressions values
+// into multiple label selectors helping with the OR condition.
+func FlattenNameSpaceSelector(nsSelector *metav1.LabelSelector) []metav1.LabelSelector {
+	//egress section
+	// for _, egressRule := range nsSelector {
+	// 	log.Logf("%+v", egressRule)
+	// }
+	if nsSelector == nil {
+		return []metav1.LabelSelector{*nsSelector}
```
Comment (JungukCho) [inline npm/parseSelector.go:148]: Not sure why it includes *nsSelector. Any reasons? Will be the same to just return nil or empty slices.

File: npm/parseSelector.go [hunk 10]
```diff
@@ -121,73 +121,170 @@ func sortSelector(selector *metav1.LabelSelector) {
...
+	}
+
+	if len(nsSelector.MatchExpressions) == 0 {
+		return []metav1.LabelSelector{*nsSelector}
```
Comment (JungukCho) [inline npm/parseSelector.go:152]: Is it different to just return nsSelector?

File: npm/parseSelector.go [hunk 11]
```diff
@@ -121,73 +121,170 @@ func sortSelector(selector *metav1.LabelSelector) {
...
+
+	baseSelector := &metav1.LabelSelector{
+		MatchLabels:      nsSelector.MatchLabels,
+		MatchExpressions: []metav1.LabelSelectorRequirement{},
+	}
+
+	multiValuePresent := false
+	multiValueMatchExprs := []metav1.LabelSelectorRequirement{}
+	for _, req := range nsSelector.MatchExpressions {
+		switch op := req.Operator; op {
```
Comment (JungukCho) [inline npm/parseSelector.go:163]: The code are exactly same in the first two cases  and in the last two cases. So, we can combine them. May be useful to use if instead switch.

File: npm/parseSelector.go (hunk 8 above)
Comment (vakalapa) [reply npm/parseSelector.go:127]: rtrStr = returnString. Corrected the variableName

File: npm/networkPolicyController.go (hunk 5 above)
Comment (vakalapa) [reply npm/networkPolicyController.go:371]: The reason I am not incrementing or decrementing a List count on creation is that, unless a network policy uses a IPSet, it is not considered be referred by others. So i chose to expose the IpSetReferIncOrDec to NetPol package, to increment the refer only when a NetworkPolicy uses this IPset, same when we delete a network policy is when we decrement the value

File: npm/networkPolicyController.go (hunk 6 above)
Comment (vakalapa) [reply npm/networkPolicyController.go:361]: As i explained above, this should not be used in createList, but only when netowrkPolicy or a 2nd level List uses an IPset

File: npm/parseSelector.go (hunk 7 above)
Comment (vakalapa) [reply npm/parseSelector.go:17]: this was earlier used in creating the single podselector for processed NPMap. this is just a lint change, and i do not want to touch this part in this PR, as the scope is different.

File: npm/parseSelector.go (hunk 9 above)
Comment (vakalapa) [reply npm/parseSelector.go:148]: The way we designed, FlattenNSSelector will return a list of selectors, so it can be looped over, thats why for an nill or empty selector, i am just returning the original selector as a list.

File: npm/parseSelector.go (hunk 10 above)
Comment (vakalapa) [reply npm/parseSelector.go:152]: See previous comment.

File: npm/parseSelector.go (hunk 8 above)
Comment (rbtr) [reply npm/parseSelector.go]: @vakalapa `returnString` is not clearer; in Go it is better to name the variable by what it holds, not by its type or use - "~returnString~ is the ... (`IPSetName` maybe?)"
maybe that is what @JungukCho was getting at

File: npm/parseSelector.go (hunk 9 above)
Comment (JungukCho) [reply npm/parseSelector.go:148]: Is "[]metav1.LabelSelector{*nsSelector} =>  []metav1.LabelSelector{}" different if nsSelector is nil?

File: npm/parseSelector_test.go [hunk 12]
```diff
@@ -488,41 +446,113 @@ func TestParseSelector(t *testing.T) {
...
+				},
+			},
+		},
+		MatchLabels: map[string]string{
+			"c": "d",
+			"a": "b",
+		},
+	}
+
+	testSelectors := FlattenNameSpaceSelector(firstSelector)
```
Comment (JungukCho) [inline npm/parseSelector_test.go:511]: Will be very helpful if there are comparison between expected values and testSelectors to easily understand what expected values are.

File: npm/parseSelector_test.go [hunk 13]
```diff
@@ -488,41 +446,113 @@ func TestParseSelector(t *testing.T) {
...
+				},
+			},
+		},
+		MatchLabels: map[string]string{
+			"c": "d",
+			"a": "b",
+		},
+	}
+
+	testSelectors = FlattenNameSpaceSelector(secondSelector)
```
Comment (JungukCho) [inline npm/parseSelector_test.go:554]: Will be very helpful if there are comparison between expected values and testSelectors to easily understand what expected values are from FlattenNameSpaceSelector and improve debugging capability of UTs.

File: npm/parseSelector.go [hunk 14]
```diff
@@ -121,73 +121,162 @@ func sortSelector(selector *metav1.LabelSelector) {
...
+	}
+
+	multiValuePresent := false
+	multiValueMatchExprs := []metav1.LabelSelectorRequirement{}
+	for _, req := range nsSelector.MatchExpressions {
+
+		if (req.Operator == metav1.LabelSelectorOpIn) || (req.Operator == metav1.LabelSelectorOpNotIn) {
+			if len(req.Values) == 1 {
+				baseSelector.MatchExpressions = append(baseSelector.MatchExpressions, req)
+				continue
```
Comment (JungukCho) [inline npm/parseSelector.go:167]: May be better readability to use "else" instead of "continue"

File: npm/parseSelector.go [hunk 15]
```diff
@@ -121,73 +121,162 @@ func sortSelector(selector *metav1.LabelSelector) {
...
+		}
+	}
+
+	// If there are no multiValue NS selector match expressions
+	// return the original NsSelector
+	if !multiValuePresent {
+		return []metav1.LabelSelector{*nsSelector}
+	}
+
+	// Now use the baseSelector and loop over multiValueMatchExprs to create all
```
Comment (JungukCho) [inline npm/parseSelector.go:185]: Highly suggest moving codes from 185-192 to "zipMatchExprs". It will improve readability. Now, I need to track for loop in 190 with zipMatchExprs function which uses returned values as input again.

File: npm/parseSelector.go [hunk 16]
```diff
@@ -121,73 +121,162 @@ func sortSelector(selector *metav1.LabelSelector) {
...
+	for _, req := range multiValueMatchExprs {
+		flatNsSelectors = zipMatchExprs(flatNsSelectors, req)
+	}
+
+	return flatNsSelectors
+}
+
+func zipMatchExprs(baseSelectors []metav1.LabelSelector, matchExpr metav1.LabelSelectorRequirement) []metav1.LabelSelector {
+	zippedLabelSelectors := []metav1.LabelSelector{}
+	for _, selector := range baseSelectors {
```
Comment (JungukCho) [inline npm/parseSelector.go:199]: Can you check whether this work or not when nsSelector does not have "MatchLabels" fields and have only multiple values in "MatchExpressions" field? Seems it does not iterate for loop at all. Also add this case in UTs to confirm this function.

File: npm/parseSelector.go (hunk 15 above)
Comment (JungukCho) [reply npm/parseSelector.go:185]: Or moving codes from "zipMatchExprs" to here. The function name of "zipMatchExprs" is not very intuitive at least to me.  I think "FlattenNameSpaceSelector" function has a good name and combining all codes in the function is good to understand and not deep.

File: npm/translatePolicy.go [hunk 17]
```diff
@@ -113,17 +114,44 @@ func craftPartialIptEntrySpecFromOpsAndLabels(ns string, ops, labels []string, s
...
+func craftPartialIptEntrySpecFromSelector(ns string, selector *metav1.LabelSelector, srcOrDstFlag string, isNamespaceSelector bool) ([]string, []string, map[string][]string) {
+	// parse the sector into labels and maps of multiVal match Exprs
+	labelsWithOps, nsLabelListKVs := parseSelector(selector)
 	ops, labels := GetOperatorsAndLabels(labelsWithOps)
-	return craftPartialIptEntrySpecFromOpsAndLabels(ns, ops, labels, srcOrDstFlag, isNamespaceSelector)
+
+	valueLabels := []string{}
+	listLabelsWithMembers := make(map[string][]string)
+	labelsForSpec := labels
+	// TODO add comments very confusing for sure
```
Comment (JungukCho) [inline npm/translatePolicy.go:134]: Can you add the detailed comments in this PR? It is good to have comments now when you have contexts in details and helps reviwers to understand the codes as well.

File: npm/translatePolicy.go [hunk 18]
```diff
@@ -113,17 +114,44 @@ func craftPartialIptEntrySpecFromOpsAndLabels(ns string, ops, labels []string, s
...
 	}
 
 	return spec
 }
 
 // craftPartialIptEntrySpecFromSelector :- ns must be "" for namespace selectors
-func craftPartialIptEntrySpecFromSelector(ns string, selector *metav1.LabelSelector, srcOrDstFlag string, isNamespaceSelector bool) []string {
-	labelsWithOps, _, _ := parseSelector(selector)
+// TODO check all references of this func
+func craftPartialIptEntrySpecFromSelector(ns string, selector *metav1.LabelSelector, srcOrDstFlag string, isNamespaceSelector bool) ([]string, []string, map[string][]string) {
```
Comment (JungukCho) [inline npm/translatePolicy.go:126]: Will be very helpful to add comment of this function

File: npm/translatePolicy.go [hunk 19]
```diff
@@ -160,11 +195,25 @@ func craftPartialIptablesCommentFromSelector(ns string, selector *metav1.LabelSe
...
+		}
+	}
+}
+
+func translateIngress(ns string, policyName string, targetSelector metav1.LabelSelector, rules []networkingv1.NetworkPolicyIngressRule) ([]string, []string, map[string][]string, [][]string, []*iptm.IptEntry) {
 	var (
-		sets                                  []string // ipsets with type: net:hash
-		namedPorts                            []string // ipsets with type: hash:ip,port
-		lists                                 []string // ipsets with type: list:set
+		sets                                  []string            // ipsets with type: net:hash
```
Comment (JungukCho) [inline npm/translatePolicy.go:214]: nitPick: better to name netHashIPsets?

File: npm/translatePolicy.go [hunk 20]
```diff
@@ -160,11 +195,25 @@ func craftPartialIptablesCommentFromSelector(ns string, selector *metav1.LabelSe
...
+		namedPorts                            []string            // ipsets with type: hash:ip,port
+		lists                                 map[string][]string // ipsets with type: list:set
```
Comment (JungukCho) [inline npm/translatePolicy.go:216]: nitPick: better to name listIPsets?

File: npm/translatePolicy.go [hunk 21]
```diff
@@ -423,21 +473,129 @@ func translateIngress(ns string, policyName string, targetSelector metav1.LabelS
...
-				nsLabelsWithOps, _, _ := parseSelector(fromRule.NamespaceSelector)
-				_, nsLabelsWithoutOps := GetOperatorsAndLabels(nsLabelsWithOps)
-				if len(nsLabelsWithoutOps) == 1 && nsLabelsWithoutOps[0] == "" {
-					// Empty namespaceSelector. This selects all namespaces
-					nsLabelsWithoutOps[0] = util.KubeAllNamespacesFlag
-				} else {
-					for i, _ := range nsLabelsWithoutOps {
-						// Add namespaces prefix to distinguish namespace ipset lists and pod ipsets
-						nsLabelsWithoutOps[i] = "ns-" + nsLabelsWithoutOps[i]
+				for _, nsSelector := range FlattenNameSpaceSelector(fromRule.NamespaceSelector) {
```
Comment (JungukCho) [inline npm/translatePolicy.go:476]: Prefer to splitting this one line to two lines to understand what it returns. Same to all in this file.

**Comment by JungukCho** [comment]
It would be nice to leave comments to describe high-level behaviors (not very details) of your updates in this PR.

File: npm/parseSelector.go [hunk 22]
```diff
@@ -121,73 +121,162 @@ func sortSelector(selector *metav1.LabelSelector) {
...
-	vals = append(vals, sortedVals...)
 
 	for _, req := range selector.MatchExpressions {
 		var k string
 		switch op := req.Operator; op {
 		case metav1.LabelSelectorOpIn:
+			k = req.Key
+			if len(req.Values) == 1 {
+				labels = append(labels, k+":"+req.Values[0])
+				continue
```
Comment (JungukCho) [inline npm/parseSelector.go:251]: better to use else instead of continue?

File: npm/parseSelector.go [hunk 23]
```diff
@@ -121,73 +121,162 @@ func sortSelector(selector *metav1.LabelSelector) {
...
-				keys = append(keys, k)
-				vals = append(vals, v)
-				labels = append(labels, k+":"+v)
+				vals[k] = append(vals[k], v)
 			}
 		case metav1.LabelSelectorOpNotIn:
+			k = util.IptablesNotFlag + req.Key
+			if len(req.Values) == 1 {
+				labels = append(labels, k+":"+req.Values[0])
+				continue
```
Comment (JungukCho) [inline npm/parseSelector.go:263]: better to use else instead of continue?

File: npm/parseSelector.go (hunk 16 above)
Comment (vakalapa) [reply npm/parseSelector.go:199]: Added the UT to confirm and it works as expected. Without MatchLabel it just loops over the matchExprs

File: npm/parseSelector_test.go (hunk 13 above)
Comment (vakalapa) [reply npm/parseSelector_test.go:554]: Very good idea, i added more deepEqual checks to make sure the output is as expected.

File: npm/parseSelector_test.go (hunk 12 above)
Comment (vakalapa) [reply npm/parseSelector_test.go:511]: Agreed, added comparison values

File: npm/parseSelector.go (hunk 15 above)
Comment (vakalapa) [reply npm/parseSelector.go:223]: I modelled it as a recursive function, where we take in a slice and expand it using each expr.

File: npm/parseSelector_test.go [hunk 24]
```diff
@@ -513,6 +513,10 @@ func TestFlattenNameSpaceSelector(t *testing.T) {
 		t.Errorf("TestFlattenNameSpaceSelector failed @ 1st selector length check %+v", testSelectors)
 	}
 
+	if !reflect.DeepEqual(testSelectors[0], *firstSelector) {
```
Comment (JungukCho) [inline npm/parseSelector_test.go:516]: There are multiple time to check length of returned values from FlattenNameSpaceSelector and deepEqual in this function and TestFlattenNameSpaceSelectorWoMatchLabels function. So, I think it is better to make one helper function to check them together.

File: npm/parseSelector_test.go [hunk 25]
```diff
@@ -555,4 +559,389 @@ func TestFlattenNameSpaceSelector(t *testing.T) {
...
+		},
+	}
+
+	if !reflect.DeepEqual(expectedSelectors, testSelectors) {
+		t.Errorf("TestFlattenNameSpaceSelector failed @ 2nd selector deepEqual check.\n Expected: %+v \n Actual: %+v", expectedSelectors, testSelectors)
+	}
+}
+
+func TestFlattenNameSpaceSelectorWoMatchLabels(t *testing.T) {
+	firstSelector := &metav1.LabelSelector{
```
Comment (JungukCho) [inline npm/parseSelector_test.go:844]: Same here. You can make it concise and better readability by using variables to avoid unnecessary repeat.

File: npm/networkPolicyController.go (hunk 6 above)
Comment (JungukCho) [reply npm/networkPolicyController.go:361]: In this case, I prefer to having another function which functions "CreateList and IPSetReferIncOrDec". It will be easy to remember and help to encapsulate details. 
I assume that your next PR is about those reference counts, we can revisit this in the PR.

File: npm/ipsm/ipsm.go [hunk 26]
```diff
@@ -462,6 +462,33 @@ func (ipsMgr *IpsetManager) DeleteFromSet(setName, ip, podKey string) error {
 	return nil
 }
 
+// Clean removes all the empty sets & lists under the namespace.
```
Comment (JungukCho) [inline npm/ipsm/ipsm.go]: When do we use this function?

File: npm/ipsm/ipsm.go [hunk 27]
```diff
@@ -87,7 +87,7 @@ func (ipsMgr *IpsetManager) Exists(listName string, setName string, kind string)
 	return true
 }
 
-// IpSetReferIncOrDec checks if an element exists in setMap/listMap and then increases or decreases tis refer count.
+// IpSetReferIncOrDec checks if an element exists in setMap/listMap and then increases or decreases this referCount.
 func (ipsMgr *IpsetManager) IpSetReferIncOrDec(ipsetName string, kind string, countOperation ReferCountOperation) {
```
Comment (JungukCho) [inline npm/ipsm/ipsm.go:91]: Do you have plan to extend this function for other types in addition to "list" as well?
In this PR, kind is always "list".

File: npm/networkPolicyController.go (hunk 6 above)
Comment (vakalapa) [reply npm/networkPolicyController.go:361]: I agree, in refer count work item, we can look at different scenarios and better organize these functions.

File: npm/ipsm/ipsm.go (hunk 27 above)
Comment (vakalapa) [reply npm/ipsm/ipsm.go:91]: Yes in the next work item(refer coutn) we will extend support of this IncDec to SetMap where we will track the use of iptable rules and their references to a given ipset. we will track both kinds of ipsets.

File: npm/ipsm/ipsm.go (hunk 4 above)
Comment (vakalapa) [reply npm/ipsm/ipsm.go:91]: We will need to expose this to network policy controller as netpol is the one referencing out ipsetmaps

File: npm/parseSelector.go [hunk 28]
```diff
@@ -140,10 +140,41 @@ func HashSelector(selector *metav1.LabelSelector) string {
 // flattenNameSpaceSelector will help flatten multiple NameSpace selector match Expressions values
 // into multiple label selectors helping with the OR condition.
 func FlattenNameSpaceSelector(nsSelector *metav1.LabelSelector) []metav1.LabelSelector {
-	//egress section
-	// for _, egressRule := range nsSelector {
-	// 	log.Logf("%+v", egressRule)
-	// }
+	/*
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply npm/parseSelector.go:143]: This comes back to what we discussed in our meeting today, having a visual representation of what the YAML looks like and what the transformation looks like will help other developers to quickly understand. I would like to keep this as is if its ok.

File: npm/parseSelector_test.go (hunk 24 above)
Comment (vakalapa) [reply npm/parseSelector_test.go:516]: Makes sense.

File: npm/ipsm/ipsm.go (hunk 26 above)
Comment (vakalapa) [reply npm/ipsm/ipsm.go]: Clean() function is not being currently, this is for future work items.

File: npm/parseSelector_test.go [hunk 29]
```diff
@@ -555,4 +559,389 @@ func TestFlattenNameSpaceSelector(t *testing.T) {
 	if len(testSelectors) != 8 {
 		t.Errorf("TestFlattenNameSpaceSelector failed @ 2nd selector length check %+v", testSelectors)
 	}
+
+	expectedSelectors := []metav1.LabelSelector{
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply npm/parseSelector_test.go:563]: added a common matchLabel to help with readability

File: npm/parseSelector.go (hunk 9 above)
Comment (JungukCho) [reply npm/parseSelector.go:179]: Luckily, no one from translate.go calls this FlattenNameSpaceSelector function with **nil nsSelector** value in current NPM. 
However, we are not sure who calls this function in the future. So, it is good to check nil value at the first place.

However, in that case, it should be below code if "nsSelector" is nil. 
Otherwise, it will be panic. We cannot dereference nil value.
```go
return []metav1.LabelSelector{}
```

Comment (JungukCho) [reply npm/parseSelector.go]: I think this is a good lesson. We should add various and edge cases in UTs. 
Suggest adding UT for nil test for this function.

File: npm/parseSelector.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (JungukCho) [reply npm/parseSelector.go:143]: To be clear, I suggested putting comprehensive information (e.g., pods yaml and netpol yaml which shows the problem in our NPM) to reproduce the problem, which helps to understand big picture at the first. T
his code is very straightforward. So, I don't think we need this, but if you want to keep it, I don't have strong opinion.

File: npm/ipsm/ipsm.go (hunk 26 above)
Comment (JungukCho) [reply npm/ipsm/ipsm.go]: I would suggest not to add function which is not used in this PR. I think this is a good practice. 
There are multiple reasons. 
1. Reviewer do not need to spend time. 
2. Add the function with the PR for your future work items. So, it will be well aligned and easy to understand PR. 
3. There are possibility that this function will not be used in the future. At that time, this function just stays for long time. 

I think you have future work item to use this function. So, add this function in the PR.

**Review by JungukCho — APPROVED** [review]
Thank you for applying comments. Look forward to your next PR.

**Review by JungukCho — APPROVED** [review]

**Comment by vakalapa** [comment]
Merging, we got a green in Cyclonus.
